### DISCUSSION TITLE
Switch between source and header in C++

### DISCUSSION THREAD

**Discussion by shahamran** [helix-editor/helix#8125] [Q&A] [answered]
Is there a way to have Helix send `clangd`'s `textDocument/switchSourceHeader` request?

Reference: <https://clangd.llvm.org/extensions#switch-between-sourceheader>

**Comment by the-mikedavis** [comment, marked as answer]
We don't support language-server-specific extensions to the protocol. This should be covered by a plugin once the plugin system exists

**Reply by shahamran** [reply]
I see, makes sense.

Can you please explain what's the `:lsp-workspace-command` command?
Is it used to send a standard LSP request?

Thanks!

**Reply by the-mikedavis** [reply]
That is a part of the spec, see https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_executeCommand

**Comment by 3541** [comment]
To close the loop, this is now possible via the Steel PR: https://github.com/helix-editor/helix/pull/8675#issuecomment-2310102922.

**Comment by objectiveryan** [comment]
I _really_ needed this feature, so I found a workaround.

With https://github.com/helix-editor/helix/pull/4709 (which has already been merged into `master` but isn't yet in an official release) and a helper shell script I called `altname` in my PATH, I use this key binding:

```
# Go to "other" file (header <--> source)
g = { o = "@\"%P|altfile<ret>gf<C-o>uu<C-i>" }
```

The `altfile` shell script:
```shell
#!/bin/bash

read f
f="$(basename "$f")"

if [[ $f == *.h ]]; then
  echo "${f%.h}.cc"
elif [[ $f == *.cc ]]; then
  echo "${f%.cc}.h"
else
  echo "$f"
fi
```

It doesn't handle non-.h/.cc files well, but it seems to work reliably when used correctly.

**Reply by pierrelgol** [reply]
Can't you use bear to generate the compile_commands.json ? When I use `bear -- make`, it creates a compile_commands.json in my root directory, clangd is then able to provide out of the box in Helix, the goto header file, I don't know if that's what you actually wanted to do. But if I'm understanding this conversation, this works very reliably and very well.

**Reply by objectiveryan** [reply]
I'm new to Helix so maybe I overcomplicated it.

Is it possible to use clangd to switch between source and header, before the plugin system is ready?

**Reply by 3541** [reply]
> I'm new to Helix so maybe I overcomplicated it.
> 
> 
> 
> Is it possible to use clangd to switch between source and header, before the plugin system is ready?

If there is a way to invoke the clangd request via the CLI, you could combine the two approaches—bind a key that invokes a script.

**Reply by pierrelgol** [reply]
@objectiveryan Yes as mentionned in my first response, you just have to use `bear` to generate the compile_commands.json, this is not a helix issue, it's a clangd issue, it needs a compile_commands.json, to index where each tokens are, which in turns provide the goto feature.

**Reply by objectiveryan** [reply]
@pierrelgol I have a compile_commands.json, and I can jump to declarations and definitions, but I haven’t found a way to easily jump back and forth between a header and its source file. Is there a way?

**Reply by pierrelgol** [reply]
you can do `gd` to go from a header file to the declaration of something.

**Reply by objectiveryan** [reply]
That's not the problem here. As mentioned above, I can already jump to declarations and definitions.

I'm talking about just jumping to the other file, for example when the declaration or definition doesn't exist yet, like when I just declared a function in the header and now I need to implement it (or vice versa). Or when I just modified the signature of a function, and now clangd *won't* jump to the implementation anymore because the header and source files disagree, and I need to go fix it.

This a common-enough need that clangd reluctantly [extends the LSP protocol](https://clangd.llvm.org/extensions#switch-between-sourceheader) for it. Helix will support that with plugins, but until that's ready, the hacky workaround I posted above has really helped me and maybe it can help others too.

**Reply by pierrelgol** [reply]
Do you use tabs ? I usually use the file picker to switch when something doesn't exist yet. But yeah instead of making their tool better I guess they prefer to extend the protocol and defer the wok on the editors.
